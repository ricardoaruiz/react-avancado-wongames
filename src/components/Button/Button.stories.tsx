import { Story, Meta } from '@storybook/react/types-6-0'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'
import Button, { ButtonProps } from './Button'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    size: {
      type: 'select',
      options: ['small', 'medium', 'large']
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  children: 'By now',
  size: 'medium'
}

export const WithIcon: Story<ButtonProps> = (args) => <Button {...args} />

WithIcon.args = {
  children: 'By now',
  size: 'small',
  icon: <AddShoppingCart />
}

export const Minimal: Story<ButtonProps> = (args) => <Button {...args} />

Minimal.args = {
  children: 'By now',
  size: 'small',
  icon: <AddShoppingCart />,
  minimal: true
}

export const AsLink: Story<ButtonProps> = (args) => <Button {...args} />

AsLink.args = {
  children: 'By now',
  size: 'large',
  as: 'a',
  href: '/link'
}
