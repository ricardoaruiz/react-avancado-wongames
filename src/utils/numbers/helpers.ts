export const toCurrencyString = (value: number) =>
  Intl.NumberFormat('pt-BR', {
    style: 'currency',
    currency: 'BRL'
  }).format(value)
