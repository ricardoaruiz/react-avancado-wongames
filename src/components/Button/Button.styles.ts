import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from './Button'

type WrapperProps = Pick<ButtonProps, 'size' | 'fullWidth'>

const wrapperModifiers = {
  sizes: {
    small: (theme: DefaultTheme) => css`
      height: 3rem;
      font-size: ${theme.font.sizes.xsmall};

      & svg {
        width: ${theme.font.sizes.small};
      }
    `,
    medium: (theme: DefaultTheme) => css`
      height: 4rem;
      font-size: ${theme.font.sizes.small};
      padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};

      & svg {
        width: ${theme.font.sizes.large};
      }
    `,
    large: (theme: DefaultTheme) => css`
      height: 5rem;
      font-size: ${theme.font.sizes.medium};
      padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};

      & svg {
        width: ${theme.font.sizes.xlarge};
      }
    `,
    fullWidth: () => css`
      width: 100%;
    `
  }
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullWidth }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;

    color: ${theme.colors.white};
    background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
    border: none;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    font-weight: ${theme.font.bold};
    outline: none;
    cursor: pointer;
    text-decoration: none;

    &:hover {
      background: linear-gradient(180deg, #e35565 0%, #d958a6 100%),
        linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
    }

    & svg {
      margin-right: ${theme.spacings.xxsmall};
    }

    /* Modifiers */
    ${!!size && wrapperModifiers.sizes[size](theme)}
    ${fullWidth && wrapperModifiers.sizes.fullWidth()};
  `}
`
