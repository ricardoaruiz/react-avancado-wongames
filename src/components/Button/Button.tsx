import React, { ButtonHTMLAttributes, AnchorHTMLAttributes } from 'react'
import * as S from './Button.styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  icon?: React.ReactNode
  as?: React.ElementType
} & ButtonTypes

export const Button = ({
  children,
  size = 'medium',
  fullWidth = false,
  icon,
  ...props
}: ButtonProps) => {
  return (
    <S.Wrapper size={size} fullWidth={fullWidth} {...props}>
      {!!icon && icon}
      {!!children && <span>{children}</span>}
    </S.Wrapper>
  )
}

export default Button
