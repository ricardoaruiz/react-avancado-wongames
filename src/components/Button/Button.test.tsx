import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from './Button'

describe('<Button />', () => {
  it('should render a medium button by default - snapshot', () => {
    const { container } = renderWithTheme(<Button>Buy now</Button>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a medium button by default', () => {
    renderWithTheme(<Button>Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '4rem',
      'font-size': '1.4rem',
      padding: '0.8rem 3.2rem'
    })
  })

  it('should render a small button when size small is passed', () => {
    renderWithTheme(<Button size="small">Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render a large button when size large is passed', () => {
    renderWithTheme(<Button size="large">Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '5rem',
      'font-size': '1.6rem',
      padding: '0.8rem 4.8rem'
    })
  })

  it('should render a full width button when fullWidth is passed', () => {
    renderWithTheme(<Button fullWidth>Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render a button with icon', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Buy now</Button>
    )

    expect(screen.getByRole('button', { name: /Buy now/i })).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })
})
