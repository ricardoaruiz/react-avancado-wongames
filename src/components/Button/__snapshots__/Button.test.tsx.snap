// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a medium button by default - snapshot 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FAFAFA;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  font-weight: 600;
  outline: none;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 100%),linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
}

.c0 svg {
  margin-right: 0.8rem;
}

.c0 svg {
  width: 1.8rem;
}

<button
  class="c0"
>
  <span>
    Buy now
  </span>
</button>
`;
